pkg/vmprovider/providers/vsphere2/virtualmachine/configspec_test.go

```diff
@@ -108,6 +111,37 @@ var _ = Describe("CreateConfigSpec", func() {
 			_, ok := dSpec.Device.(*vimtypes.VirtualE1000)
 			Expect(ok).To(BeTrue())
 		})
+
+		When("Image firmware is empty", func() {
+			BeforeEach(func() {
+				vmImageStatus = &vmopv1.VirtualMachineImageStatus{}
+			})
+
+			It("config spec has the firmware from the class", func() {
+				Expect(configSpec.Firmware).ToNot(Equal(vmImageStatus.Firmware))
+				Expect(configSpec.Firmware).To(Equal(classConfigSpec.Firmware))
+			})
+		})
+
+		When("vm has a valid firmware override annotation", func() {
+			BeforeEach(func() {
+				vm.Annotations[constants.FirmwareOverrideAnnotation] = "efi"
+			})
+
+			It("config spec has overridden firmware annotation", func() {
+				Expect(configSpec.Firmware).To(Equal(vm.Annotations[constants.FirmwareOverrideAnnotation]))
+			})
+		})
+
+		When("vm has an invalid firmware override annotation", func() {
+			BeforeEach(func() {
+				vm.Annotations[constants.FirmwareOverrideAnnotation] = "foo"
+			})
+
+			It("config spec doesn't have the invalid val", func() {
+				Expect(configSpec.Firmware).ToNot(Equal(vm.Annotations[constants.FirmwareOverrideAnnotation]))
```

Comment:
Just to confirm, if vmi doesn't have firmware specified, and annotation has invalid val. ConfigSpec firmware will be empty val?

Reply:
So it goes like,
If 
- annotation has invalid val or doesn't exist then
- vmi doesn't have firmware specified and 
- vmclass doesn't have firmware specified 

then configSpec will be empty when passed on to vpxd.

Reply:
so annotation takes precedence to image, image take precedence to vmclass?

Reply:
annotation is an override to everything. yes.